server/src/main/java/com/paypal/selion/proxy/SeLionSauceProxy.java

```diff
@@ -22,72 +22,248 @@
+ * This customized version of {@link DefaultRemoteProxy} basically helps redirect all traffic to the SauceLabs cloud.
  */
-public class SeLionSauceProxy extends BaseRemoteProxy {
+public class SeLionSauceProxy extends DefaultRemoteProxy {
 
     private static final SeLionGridLogger LOGGER = SeLionGridLogger.getLogger(SeLionSauceProxy.class);
+
+    /** {@link SauceConfigReader} instruction to use the provided username value also for the parent-tunnel, if omitted */
+    public static final String DEFAULT_PARENT_TUNNEL_TO_USERNAME_TOKEN = "$username";
+    /** Sauce labs cloud default idle timeout is 90 seconds */
+    public static final int DEFAULT_IDLE_TIMEOUT = 90000;
+    /** Overrides value in DefaultRemoteProxy */
+    public static final int DEFAULT_POLLING_INTERVAL = 60000;
+    /** Overrides value in DefaultRemoteProxy */
+    public static final int DEFAULT_DOWN_POLLING_LIMIT = 3;
+
     private final SauceLabsRestApi sauceApi;
+    private final String defaultParentTunnel;
+    private final String defaultTunnelIdentifier;
     private final Lock accessLock = new ReentrantLock();
+    private final SauceConfigReader sauceConfigReader;
+    private Thread pollingThread;
+
+    /*
+     * Self Healing. Polling configuration
+     */
+    private volatile int pollingInterval;
+    private volatile int downPollingLimit;
+
+    /*
+     * Self Healing part. Polls the remote, and marks it down if it cannot be reached twice in a row.
+     */
+    private volatile boolean down;
+    private volatile boolean poll;
+
+    /**
+     * Sauce labs DesiredCapabilities that {@link SeLionSauceProxy} is concerned with
+     */
+    public static class SauceLabsCapability {
+        public static final String SAUCE_USER_NAME_CAPABILITY = "sauceUserName";
+        public static final String SAUCE_API_KEY_CAPABILITY = "sauceApiKey";
+        public static final String USERNAME_CAPABILITY = "username";
+        public static final String ACCESS_KEY_CAPABILITY = "accessKey";
+        public static final String PARENT_TUNNEL_CAPABILITY = "parent-tunnel";
+        public static final String TUNNEL_IDENTIFIER_CAPABILITY = "tunnelIdentifier";
+        public static final String IDLE_TIMEOUT_CAPABILITY = "idleTimeout";
+    }
+
+    private static RegistrationRequest fixTimeout(RegistrationRequest request) {
+        // make sure there is a timeout set for this proxy
+        int timeout = request.getConfigAsInt(RegistrationRequest.TIME_OUT, DEFAULT_IDLE_TIMEOUT);
+        request.getConfiguration().put(RegistrationRequest.TIME_OUT, timeout);
+        return request;
+    }
 
     public SeLionSauceProxy(RegistrationRequest request, Registry registry) {
-        super(request, registry);
+        super(fixTimeout(request), registry);
 
-        if (! SauceServlet.PROXY_ID.equals(remoteHost.toExternalForm())) {
+        down = false;
+        poll = true;
```

Comment:
Do we want to allow being able to turn off polling by specifying a polling interval of "0"